File: pt6/1282.group-the-people-given-the-group-size-they-belong-to.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        Map<Integer, List<Integer>> buckets = new HashMap<>();
        List<List<Integer>> ans = new ArrayList<>();

        for (int person = 0; person < groupSizes.length; person++) {
            int size = groupSizes[person];
            buckets.computeIfAbsent(size, key -> new ArrayList<>()).add(person);

            if (buckets.get(size).size() == size) {
                ans.add(buckets.get(size));
                buckets.put(size, new ArrayList<>());
            }
        }

        return ans;
    }
}

/*
Explanation

Bucket people by required group size. Whenever a bucket reaches that size, it
forms a complete group and is added to the answer.

HashMap<Integer, List<Integer>> is the right Java data structure because each
group size owns one current partial group. People with the same size are
interchangeable, and the problem guarantees a valid partition.

Edge cases: group size 1 flushes immediately; many groups may share the same
size; any valid output order is accepted.

Time complexity: O(n).
Space complexity: O(n).
*/

File: pt6/1282.group-the-people-given-the-group-size-they-belong-to.go
package main

func groupThePeople(groupSizes []int) [][]int {
	buckets := map[int][]int{}
	ans := [][]int{}

	for person, size := range groupSizes {
		buckets[size] = append(buckets[size], person)
		if len(buckets[size]) == size {
			ans = append(ans, buckets[size])
			buckets[size] = nil
		}
	}

	return ans
}

/*
Explanation

Bucket people by required group size. Whenever a bucket reaches that size, it
forms a complete group and is flushed to the answer.

A hash map from size to current partial group is enough because people with the
same group size are interchangeable. The problem guarantees that a valid
partition exists.

Edge cases: size 1 groups flush immediately; many groups can share one size;
the output order can be any valid order.

Time complexity: O(n).
Space complexity: O(n).
*/
